### PR TITLE
Add edit command

### PR DESCRIPTION
**Pull request by tohyuting**
Will update with assertions and logging in the next commit

Update:
Assertions and logging has been added for edit command

### CODE REVIEW

File: src/main/java/seedu/clinic/logic/commands/EditCommand.java [hunk 1]
```diff
@@ -26,66 +30,143 @@
...
+ * Edits the details of an existing supplier/warehouse in the CLI-nic app.
  */
 public class EditCommand extends Command {
 
     public static final String COMMAND_WORD = "edit";
 
-    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Edits the details of the supplier identified "
-            + "by the index number used in the displayed supplier list. "
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Edits the details of the supplier/warehouse"
+            + " identified " + "by the index number used in the displayed supplier/warehouse list. "
```
Comment (zhengweii) [inline src/main/java/seedu/clinic/logic/commands/EditCommand.java]: Not too sure if this is on the same line! But if it is, perhaps the concatenation might not be needed here!

File: src/main/java/seedu/clinic/logic/commands/EditCommand.java [hunk 2]
```diff
@@ -26,66 +30,143 @@
...
+            + "Note that either si or wi (not both) has to be provided, where si is the supplier list index"
+            + " and wi is the warehouse list index. In addition, a supplier should not have an address"
+            + " prefix entered while a warehouse should not have an email prefix entered. ";
 
     public static final String MESSAGE_EDIT_SUPPLIER_SUCCESS = "Edited Supplier: %1$s";
+    public static final String MESSAGE_EDIT_WAREHOUSE_SUCCESS = "Edited Warehouse: %1$s";
     public static final String MESSAGE_NOT_EDITED = "At least one field to edit must be provided.";
-    public static final String MESSAGE_DUPLICATE_SUPPLIER = "This supplier already exists in the remark book.";
+    public static final String MESSAGE_DUPLICATE_SUPPLIER = "This supplier already"
+            + " exists in the remark book.";
```
Comment (zhengweii) [inline src/main/java/seedu/clinic/logic/commands/EditCommand.java]: Can I clarify what is the remark book??

File: src/main/java/seedu/clinic/logic/commands/EditCommand.java [hunk 3]
```diff
@@ -26,66 +30,143 @@
...
+            + " index, i.e. either wi/INDEX or si/INDEX";
+    public static final String MESSAGE_NO_PREFIX = "Please enter at least one type of"
+            + " index, i.e. either wi/INDEX or si/INDEX";
+    public static final String MESSAGE_INVALID_COMMAND_FORMAT = "Invalid command format! %1$s \n"
+            + MESSAGE_USAGE;
+    public static final String MESSAGE_INVALID_PREFIX = "You used an invalid prefix!";
+    public static final String MESSAGE_SUPPLIER_PREFIX_NOT_ALLOWED = "Supplier prefix (s/) not allowed "
+            + "when editing warehouses.";
+    public static final String MESSAGE_WAREHOUSE_PREFIX_NOT_ALLOWED = "Warehouse prefix (w/) not allowed "
+            + "when editing suppliers.";
```
Comment (zhengweii) [inline src/main/java/seedu/clinic/logic/commands/EditCommand.java:78]: Oops! Lines 76 and 78 are swapped!

File: src/main/java/seedu/clinic/logic/commands/EditCommand.java [hunk 4]
```diff
@@ -212,67 +309,97 @@ public boolean equals(Object other) {
...
+        private Email email;
 
-        public EditWarehouseDescriptor() {}
+        public EditSupplierDescriptor() {
+            super();
+        }
 
         /**
          * Copy constructor.
          * A defensive copy of {@code products} is used internally.
```
Comment (zhengweii) [inline src/main/java/seedu/clinic/logic/commands/EditCommand.java]: Not too sure if this is still relevant!

File: src/main/java/seedu/clinic/logic/commands/EditCommand.java [hunk 5]
```diff
@@ -212,67 +309,97 @@ public boolean equals(Object other) {
...
+
+        public EditWarehouseDescriptor() {
+            super();
         }
 
-        public Optional<Phone> getPhone() {
-            return Optional.ofNullable(phone);
+        /**
+         * Copy constructor.
+         * A defensive copy of {@code products} is used internally.
```
Comment (zhengweii) [inline src/main/java/seedu/clinic/logic/commands/EditCommand.java]: If the above comment is relevant!

File: src/main/java/seedu/clinic/logic/parser/EditCommandParser.java [hunk 6]
```diff
@@ -33,50 +45,164 @@
...
+            Optional<Index> supplierIndex = Optional.empty();
+            try {
+                Index index = ParserUtil.parseIndex(argMultimap.getValue(PREFIX_SUPPLIER_INDEX).get());
+                supplierIndex = Optional.of(index);
+            } catch (ParseException pe) {
+                String indexValue = argMultimap.getValue(PREFIX_SUPPLIER_INDEX).get();
+                if (indexValue.contains("/")) {
+                    throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT,
+                            EditCommand.MESSAGE_INVALID_PREFIX));
+                } else {
```
Comment (zhengweii) [inline src/main/java/seedu/clinic/logic/parser/EditCommandParser.java]: Perhaps the else clause could be removed here as well!

File: src/main/java/seedu/clinic/logic/parser/EditCommandParser.java [hunk 7]
```diff
@@ -33,50 +45,164 @@
...
+            Index index;
+
+            try {
+                index = ParserUtil.parseIndex(argMultimap.getValue(PREFIX_WAREHOUSE_INDEX).get());
+            } catch (ParseException pe) {
+                String indexValue = argMultimap.getValue(PREFIX_WAREHOUSE_INDEX).get();
+                if (indexValue.contains("/")) {
+                    throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT,
+                            EditCommand.MESSAGE_INVALID_PREFIX));
+                } else {
```
Comment (zhengweii) [inline src/main/java/seedu/clinic/logic/parser/EditCommandParser.java]: Perhaps the clause can be removed here as well to make it more prominent!

File: src/main/java/seedu/clinic/logic/commands/EditCommand.java (hunk 4 above)
Comment (tohyuting) [reply src/main/java/seedu/clinic/logic/commands/EditCommand.java]: Thanks for pointing that out! Will be removing that in the next commit.

File: src/main/java/seedu/clinic/logic/parser/EditCommandParser.java (hunk 7 above)
Comment (tohyuting) [reply src/main/java/seedu/clinic/logic/parser/EditCommandParser.java]: Thanks for pointing this out, will fix it in the next commit along with removal of above clause as well!

File: src/main/java/seedu/clinic/logic/commands/EditCommand.java (hunk 1 above)
Comment (tohyuting) [reply src/main/java/seedu/clinic/logic/commands/EditCommand.java]: Thanks for pointing this out, will fix it in the next commit!

File: src/main/java/seedu/clinic/logic/commands/EditCommand.java (hunk 2 above)
Comment (tohyuting) [reply src/main/java/seedu/clinic/logic/commands/EditCommand.java]: Oh no, this should be clinic instead. Thanks for pointing this out!

File: src/main/java/seedu/clinic/logic/commands/EditCommand.java (hunk 3 above)
Comment (tohyuting) [reply src/main/java/seedu/clinic/logic/commands/EditCommand.java:78]: Oh the error message will be shown when a user tries to use a supplier prefix, s/ when editing warehouses, I think that the output message is correct though? 🤔

**Review by Criss-Wang — APPROVED** [review]
LGTM!

**Review by jeffreytjs — APPROVED** [review]
LGTM

**Pull request merged**